chain/network/src/raw.rs

```diff
@@ -0,0 +1,345 @@
+    pub async fn connect(
+        addr: SocketAddr,
+        peer_id: PeerId,
+        my_protocol_version: Option<ProtocolVersion>,
+        chain_id: &str,
+        genesis_hash: CryptoHash,
+        head_height: BlockHeight,
+    ) -> Result<Self, ConnectError> {
+        let secret_key = SecretKey::from_random(KeyType::ED25519);
+        let my_peer_id = PeerId::new(secret_key.public_key());
+
+        let start = Instant::now();
+        let stream = tokio::time::timeout(Duration::from_secs(2), TcpStream::connect(addr))
+            .await
+            .map_err(Into::<io::Error>::into)??;
+        tracing::info!(
+            target: "network", "Connection to {:?}@{:?} established. latency: {:?}",
+            &peer_id, &addr, start.elapsed(),
+        );
+
+        let mut peer =
+            Self { stream, peer_id, buf: BytesMut::with_capacity(1024), secret_key, my_peer_id };
+        peer.do_handshake(
+            my_protocol_version.unwrap_or(PROTOCOL_VERSION),
+            chain_id,
+            genesis_hash,
+            head_height,
+        )
+        .await?;
+
+        Ok(peer)
+    }
+
+    async fn do_handshake(
+        &mut self,
+        protocol_version: ProtocolVersion,
+        chain_id: &str,
+        genesis_hash: CryptoHash,
+        head_height: BlockHeight,
+    ) -> Result<(), ConnectError> {
+        let handshake = PeerMessage::Handshake(Handshake {
+            protocol_version,
+            oldest_supported_version: protocol_version - 2,
+            sender_peer_id: self.my_peer_id.clone(),
+            target_peer_id: self.peer_id.clone(),
+            // we have to set this even if we have no intention of listening since otherwise
+            // the peer will drop our connection
+            sender_listen_port: Some(24567),
+            sender_chain_info: PeerChainInfoV2 {
+                genesis_id: GenesisId { chain_id: chain_id.to_string(), hash: genesis_hash },
+                height: head_height,
+                tracked_shards: vec![0],
+                archival: false,
+            },
+            partial_edge_info: PartialEdgeInfo::new(
+                &self.my_peer_id,
+                &self.peer_id,
+                1,
+                &self.secret_key,
+            ),
+        });
+
+        self.write_message(&handshake).await?;
+
+        let start = Instant::now();
+
+        let (message, timestamp) = self.recv_message().await?;
+
+        match message {
+            // TODO: maybe check the handshake for sanity
+            PeerMessage::Handshake(_) => {
+                tracing::info!(target: "network", "handshake latency: {:?}", timestamp - start);
+            }
+            PeerMessage::HandshakeFailure(_peer_info, reason) => {
+                return Err(ConnectError::HandshakeFailure(reason))
+            }
+            _ => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
```

Comment:
technically this is not a transport layer error, so perhaps add a dedicated variant to ConnectError, sth like UnexpectedMessage?